ocaml/xapi/xapi_cluster_host.ml

```diff
@@ -65,76 +46,80 @@ let call_api_function_with_alert ~__context ~msg ~cls ~obj_uuid ~body
-      call_api_function_with_alert ~__context
-        ~msg:Api_messages.cluster_host_enable_failed
-        ~cls:`Host ~obj_uuid ~body
-        ~api_func:(fun rpc session_id -> ignore
-            (Client.Client.Cluster_host.enable rpc session_id cluster_host))
-    end
-
-let create ~__context ~cluster ~host ~pif =
-  (* TODO: take network lock *)
+(* Create xapi db object for cluster_host, resync_host calls clusterd *)
+let create_internal ~__context ~cluster ~host ~pIF : API.ref_Cluster_host =
   with_clustering_lock (fun () ->
       assert_operation_host_target_is_localhost ~__context ~host;
       assert_cluster_host_can_be_created ~__context ~host;
-      assert_pif_attached_to ~host ~pif ~__context;
+      assert_pif_attached_to ~host ~pIF ~__context;
       let ref = Ref.make () in
-      let dbg = Context.string_of_task __context in
       let uuid = Uuidm.to_string (Uuidm.create `V4) in
+      let pifrec = Db.PIF.get_record ~__context ~self:pIF in
+      assert_pif_prerequisites (pIF,pifrec);
+      Db.Cluster_host.create ~__context ~ref ~uuid ~cluster ~host ~pIF ~enabled:false
+        ~current_operations:[] ~allowed_operations:[] ~other_config:[] ~joined:false;
+
+      ref
+  )
+
+(* Helper function atomically enables clusterd and joins the cluster_host *)
+let join_internal ~__context ~self =
+  with_clustering_lock (fun () ->
+      let dbg = Context.string_of_task __context in
+      let cluster = Db.Cluster_host.get_cluster ~__context ~self in
       let cluster_token = Db.Cluster.get_cluster_token ~__context ~self:cluster in
-      let pifref,pifrec = pif,(Db.PIF.get_record ~__context ~self:pif) in
-      assert_pif_prerequisites (pifref,pifrec);
+      let pifref = Db.Cluster_host.get_PIF ~__context ~self in
+      let pifrec = Db.PIF.get_record ~__context ~self:pifref in
       let ip = ip_of_pif (pifref,pifrec) in
-      let ip_list = List.map (fun cluster_host ->
-        let pref = Db.Cluster_host.get_PIF ~__context ~self:cluster_host in
+      let ip_list = List.map (fun self ->
+        let pref = Db.Cluster_host.get_PIF ~__context ~self in
         let prec = Db.PIF.get_record ~__context ~self:pref in
           ip_of_pif (pref,prec)
         ) (Db.Cluster.get_cluster_hosts ~__context ~self:cluster) in
+
+      debug "Enabling clusterd and joining cluster_host %s" (Ref.string_of self);
       Xapi_clustering.Daemon.enable ~__context;
-      let result = Cluster_client.LocalClient.join (rpc ~__context) dbg cluster_token ip ip_list in
-      match result with
+      Cluster_client.LocalClient.join (rpc ~__context) dbg cluster_token ip ip_list
+  )
+
+(* Enable cluster_host in client layer via clusterd *)
+let resync_host ~__context ~host =
+  match (find_cluster_host ~__context ~host) with
+  | None      -> () (* no clusters exist *)
+  | Some self ->    (* cluster_host and cluster exist *)
+    if not (Db.Cluster_host.get_joined ~__context ~self) then begin
+       match join_internal ~__context ~self with
```

Comment:
We probably want the alert wrapper from enable here as well, so it would be better to wrap this whole part below `Some self` with `call_api_functions_with_alert`

Reply:
For `LocalClient.join`, I assume?

Reply:
Yes, but I think you can wrap things like this:
```
let resync_host ... =
  match ...
  | None -> ...
 | Some self ->
    call_api_functions_with_alert ~__context (fun ... ->
     if ...
         join ...
     else if ....
        enable
```
If an exception is raised by either of them you will raise an alert.
Note that call_api_functions does not mean immediately call one API function, it prepares the context for it, so you can call other functions too.